cli/dcoscli/marathon/main.py

```diff
@@ -255,312 +249,568 @@ def _get_resource(resource):
     return util.load_json(sys.stdin)
 
 
-def _add(app_resource):
-    """
-    :param app_resource: optional filename for the application resource
-    :type app_resource: str
-    :returns: process return code
-    :rtype: int
-    """
-    application_resource = _get_resource(app_resource)
-
-    # Add application to marathon
-    client = marathon.create_client()
-
-    # Check that the application doesn't exist
-    app_id = client.normalize_app_id(application_resource['id'])
-
-    try:
-        client.get_app(app_id)
-    except DCOSException as e:
-        logger.exception(e)
-    else:
-        raise DCOSException("Application '{}' already exists".format(app_id))
-
-    client.add_app(application_resource)
-
-    return 0
+class MarathonSubcommand(object):
+    """Defines a method for each operation of the `dcos marathon` subcommand.
 
+    Having the operations as methods on this class improves modularity and
+    allows common dependencies to be provided to the class constructor. It
+    also makes it much easier to mock dependencies in unit tests.
 
-def _list(json_):
-    """
-    :param json_: output json if True
-    :type json_: bool
-    :returns: process return code
-    :rtype: int
+    :param resource_reader: a callable that takes a resource file path as an
+                            argument and returns the file contents, as JSON
+    :type resource_reader: collections.abc.Callable
+    :param create_marathon_client: a callable that returns an instance of
+                                   marathon.Client
+    :type create_marathon_client: collections.abc.Callable
     """
 
-    client = marathon.create_client()
-    apps = client.get_apps()
+    def __init__(self,
+                 resource_reader=_get_resource,
+                 create_marathon_client=marathon.create_client):
+        self._resource_reader = resource_reader
+        self._create_marathon_client = create_marathon_client
 
-    if json_:
-        emitter.publish(apps)
-    else:
-        deployments = client.get_deployments()
-        table = tables.app_table(apps, deployments)
-        output = six.text_type(table)
-        if output:
-            emitter.publish(output)
+    def about(self):
+        """
+        :returns: process return code
+        :rtype: int
+        """
```

Comment:
Unfortunately putting all of these methods in a class means that they all had to be indented :disappointed: 

Fortunately GitHub bold highlights the non-whitespace changes!